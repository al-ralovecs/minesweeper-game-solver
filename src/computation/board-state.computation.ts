import BoardDto from "../dto/board.dto";
import BoardStateDto from "../dto/board-state.dto";
import ActionDto from "../dto/action.dto";
import { MarginDto } from "../dto/margin.dto";
import LocationDto from '../dto/location.dto';

export default class BoardStateComputation
{
    private board: BoardDto;
    private readonly boardState: BoardStateDto;

    public constructor(height: number, width: number)
    {
        const boardState = new BoardStateDto(height, width);
        BoardStateComputation.init(boardState);

        this.boardState = boardState;
    }

    public set setBoard(board: BoardDto)
    {
        this.board = board;
    }

    public get getBoardState(): BoardStateDto
    {
        return this.boardState;
    }

    public do(): void
    {
        for (let i: number = 0; i < this.boardState.height; i++) {
            this.boardState.adjFlagsOnBoard[i] = [];

            for (let j: number = 0; j < this.boardState.width; j++) {
                this.boardState.adjFlagsOnBoard[i][j] = 0;
            }
        }

        this.boardState.actionList = [];

        for (let i: number = 0; i < this.boardState.height; i++) {
            for (let j: number = 0; j < this.boardState.width; j++) {
                const location: LocationDto = new LocationDto(i, j);
                this.boardState.flagOnBoard[i][j] = false;
                let info: number = this.board.data[i][j];
                let act: ActionDto = this.boardState.action[i][j];

                // if (typeof act === 'undefined' && (
                //     ! act.isCertainty
                //     || MoveMethodEnum.BOOK === act.getMoveMethod
                // )) {
                //     this.boardState[i][j] = undefined;
                //     act = undefined;
                // }

                if (-1 !== info) {
                    // if (9 === info) {
                    //     this.boardState.totalFlags++;
                    //     this.boardState.flagOnBoard[i][j] = true;
                    //
                    //     let m: MarginDto = new MarginDto(i, j, this.boardState.height, this.boardState.width);
                    //
                    //     for (let k: number = i + m.top; k <= i + m.bottom; k++) {
                    //         for (let l: number = j + m.left; l <= j + m.right; l++) {
                    //             this.boardState.adjFlagsOnBoard[k][m]++;
                    //         }
                    //     }
                    // }
                    if (! this.boardState.revealed[i][j]) {
                        this.boardState.addLivingWitness(location);

                        this.boardState.revealed[i][j] = true;
                        this.boardState.board[i][j] = info;

                        let m: MarginDto = new MarginDto(i, j, this.boardState.height, this.boardState.width);

                        for (let k: number = i + m.top; k <= i + m.bottom; k++) {
                            for (let l: number = j + m.left; l <= j + m.right; l++) {
                                this.boardState.adjUnrevealed[k][j]--;
                            }
                        }
                    }
                } else {
                    if (this.boardState.flagConfirmed[i][j]) {
                        this.boardState.totalFlagsConfirmed++;
                        this.boardState.totalFlags++;
                    } else {
                        this.boardState.numOfHidden++;
                    }

                    // if (typeof act !== 'undefined' && null !== act && act.isCertainty) {
                    //     if (act.getAction !== ActionEnum.FLAG) {
                    //         this.boardState.actionList.push(act);
                    //     }
                    // }
                }
            }
        }

        // let toRemove: LocationDto[];
        // this.boardState.livingWitnesses.forEach(e => {
        //     if (0 === this.countAdjacentUnrevealed(e)) {
        //         toRemove.push(e);
        //     }
        // });
        // this.boardState.removeLivingWitnesses(toRemove);

        // this.boardState.actionList.forEach(a => {
        //    this.boardState.unPlayedMoves[a.getMoveMethod]++;
        // });
    }

    private countAdjacentUnrevealed(location: LocationDto): number
    {

    }

    private static init(boardState: BoardStateDto): void
    {
        boardState.action = [];
        boardState.flagOnBoard = [];
        boardState.adjFlagsOnBoard = [];
        boardState.adjUnrevealed = [];
        boardState.flagConfirmed = [];
        boardState.revealed = [];
        boardState.board = [];

        for (let i: number = 0; i < boardState.height; i++) {
            boardState.action[i] = [];
            boardState.flagOnBoard[i] = [];
            boardState.adjFlagsOnBoard[i] = [];
            boardState.adjUnrevealed[i] = [];
            boardState.flagConfirmed[i] = [];
            boardState.revealed[i] = [];
            boardState.board[i] = [];

            for (let j: number = 0; j < boardState.width; j++) {
                boardState.action[i][j] = undefined;
                boardState.flagOnBoard[i][j] = false;
                boardState.adjFlagsOnBoard[i][j] = 0;
                boardState.flagConfirmed[i][j] = false;
                boardState.revealed[i][j] = false;
                boardState.board[i][j] = -1;

                let adjacent: number = 8;

                if (0 === i && 0 === j
                    || 0 === i && boardState.width - 1 === j
                    || boardState.height - 1 === i && 0 === j
                    || boardState.height - 1 === i && boardState.width - 1 === j
                ) {
                    adjacent = 3;
                } else if (0 === i
                    || 0 === j
                    || boardState.height - 1 === i
                    || boardState.width - 1 === j
                ) {
                    adjacent = 5;
                }

                boardState.adjUnrevealed[i][j] = adjacent;
            }
        }
    }
}
