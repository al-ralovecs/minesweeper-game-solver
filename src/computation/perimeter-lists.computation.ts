import BoardDto from '../dto/board.dto';
import PerimeterListsListDto from '../dto/perimeter-lists-list.dto';
import {MarginDto} from "../dto/margin.dto";
import {createPerimeterListAndPopulateWithDefaults} from "../routine/perimeter-list.populate-with-defaults";
import {NeighborDto} from "../dto/neighbor.dto";

export class PerimeterListsComputation {
    private readonly board: BoardDto;

    public perimeter: PerimeterListsListDto;
    public neighbor: NeighborDto[];

    public topToBottomPriority: boolean = true;

    public used: number[][];

    public listsCount: number = 0;
    public listNext: number = 0;
    public neighborNext: number = 0;

    public listStart: number[] = [];
    public listEnd: number[] = [];

    public neighborStart: number[] = [];
    public neighborEnd: number[] = [];

    public enclosed: boolean[] = [];
    public dead: boolean[] = [];

    private nonPerimeterProbability: number;

    public constructor(board: BoardDto)
    {
        this.board = board;

        this.init();
    }

    public create(): void
    {
        if (this.topToBottomPriority) {
            for (let x: number = 0; x< this.board.width; x++) {
                for (let y: number = 0; y < this.board.height; y++) {
                    this.startList(y, x);
                }
            }
        } else {
            for (let y: number = 0; y < this.board.height; y++) {
                for (let x: number = 0; x< this.board.width; x++) {
                    this.startList(y, x);
                }
            }
        }
    }

    private startList(y: number, x: number): void
    {
        if (0 <= this.used[y][x]
            || 0 <= this.board.exposed[y][x]
            || 0 === this.board.neighbors[y][x]
        ) {
            return;
        }

        this.listStart[this.listsCount] = this.listNext;
        this.enclosed[this.listsCount] = true;
        this.dead[this.listsCount] = true;
        this.neighborStart[this.listsCount] = this.neighborNext;

        this.createList(y, x);

        this.neighborEnd[this.listsCount] = this.neighborNext - 1;
        this.listEnd[this.listsCount] = this.listNext - 1;

        this.listsCount++;
    }

    private createList(y: number, x: number, endGame: boolean = false): void
    {
        this.used[y][x] = this.listNext;

        this.perimeter.add(
            this.listNext,
            createPerimeterListAndPopulateWithDefaults(y, x, this.listsCount, this.listStart[this.listsCount])
        );

        if (this.perimeter.current.index !== this.listStart[this.listsCount]) {
            this.perimeter.previous.nextListIndex = this.perimeter.current.index; // .previous()  =?= (list - 1)->next
        }

        this.listNext++;

        if (endGame) {
            this.perimeter.current.probability = this.nonPerimeterProbability;
            return;
        }

        const m: MarginDto = new MarginDto(y, x, this.board.height, this.board.width);

        for (let i: number = y + m.top; i < y + m.bottom; i++) {
            for (let j: number = x + m.left; j < x + m.right; j++) {
                if (i === y && j === x) {
                    continue;
                }

                if (0 < this.board.exposed[i][j] && 9 > this.board.exposed[i][j]) {
                    this.addNeighbor(i, j);
                    this.checkNeighbors(i, j);
                }
            }
        }
    }

    private addNeighbor(y: number, x: number): void
    {
        const next: NeighborDto = new NeighborDto();
        this.neighbor[this.listsCount] = next;



        /**
         void AddNeighbor(List * list, int x, int y)
         {
	Neighbor * next = &neighborArray[lists];
	for (; next < &neighborArray[neighborNext]; next++)
	{
		if (next->x == x && next->y == y)
			goto FOUND;										// already on the list
	}
	neighborNext++;
	next->x = x;
	next->y = y;
	next->unexposed = unexposed[y][x];
	next->needed = needed[y][x];
	next->count = 0;
FOUND:
	next->entries[next->count++] = list;
	list->neighbors[list->numNeighbors] = next;
	list->numNeighbors++;
}
         */
    }

    private checkNeighbors(y: number, x: number): void
    {

    }

    private init(): void
    {
        for (let i: number = 0; i < this.board.height; i++) {
            this.used[i] = [];

            for (let j: number = 0; j < this.board.width; j++) {
                this.used[i][j] = -1;
            }
        }
    }
}
