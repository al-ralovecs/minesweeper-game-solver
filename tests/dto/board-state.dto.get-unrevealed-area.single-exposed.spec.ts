import { disposition } from '../fixtures/state/3x3.single-exposed.json';
import BoardDto from "../../src/dto/board.dto";
import { BoardStateComputation } from "../../src/computation/board-state.computation";
import BoardStateDto from "../../src/dto/board-state.dto";
import LocationDto from "../../src/dto/location.dto";
import AreaDto from "../../src/dto/area.dto";

describe('BoardStateDto', () => {
    test('getUnrevealedArea on a disposition with a single exposed tile', () => {
        const board = new BoardDto(disposition);
        const boardStateComputation = new BoardStateComputation(board.height, board.width);

        boardStateComputation.setBoard = board;
        boardStateComputation.process();

        const boardState: BoardStateDto = boardStateComputation.getBoardState;

        const unrevealed: number = boardState.getTotalUnrevealedCount;
        expect(unrevealed).toBe(8);

        const allWitnesses: LocationDto[] = boardState.getAllLivingWitnesses;
        const expectedWitnesses = [
            new LocationDto(2, 0)
        ];
        expect(allWitnesses).toStrictEqual(expectedWitnesses);

        const allWitnessedSquares: AreaDto = boardState.getUnrevealedArea(allWitnesses);
        expect(allWitnessedSquares).toMatchSnapshot();
    });
});
