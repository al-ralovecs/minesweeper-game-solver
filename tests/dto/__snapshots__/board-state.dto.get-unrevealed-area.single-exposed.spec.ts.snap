// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardStateDto getUnrevealedArea on a disposition with a single exposed tile 1`] = `
AreaDto {
  "area": LocationSetDto {
    "data": Array [
      LocationDto {
        "x": 0,
        "y": 1,
      },
      LocationDto {
        "x": 1,
        "y": 1,
      },
      LocationDto {
        "x": 1,
        "y": 2,
      },
    ],
  },
}
`;
